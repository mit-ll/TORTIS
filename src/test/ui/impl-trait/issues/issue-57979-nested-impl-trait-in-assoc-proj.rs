// rust-lang/rust#57979 : the initial support for `impl Trait` didn't
// properly check syntax hidden behind an associated type projection.
// Here we test behavior of occurrences of `impl Trait` within an
// `impl Trait` in that context.

mod allowed {
    #![allow(nested_impl_trait)]

    pub trait Foo<T> { }
    pub trait Bar { }
    pub trait Quux { type Assoc; }
    pub fn demo(_: impl Quux<Assoc=impl Foo<impl Bar>>) { }
}

mod warned {
    #![warn(nested_impl_trait)]

    pub trait Foo<T> { }
    pub trait Bar { }
    pub trait Quux { type Assoc; }
    pub fn demo(_: impl Quux<Assoc=impl Foo<impl Bar>>) { }
    //~^ WARN nested `impl Trait` is not allowed
    //~| WARN will become a hard error in a future release!
}

mod denied {
    #![deny(nested_impl_trait)]

    pub trait Foo<T> { }
    pub trait Bar { }
    pub trait Quux { type Assoc; }
    pub fn demo(_: impl Quux<Assoc=impl Foo<impl Bar>>) { }
    //~^ ERROR nested `impl Trait` is not allowed
    //~| WARN will become a hard error in a future release!
}

fn main() { }
