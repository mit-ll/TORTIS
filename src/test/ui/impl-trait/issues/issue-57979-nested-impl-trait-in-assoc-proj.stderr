warning: nested `impl Trait` is not allowed
  --> $DIR/issue-57979-nested-impl-trait-in-assoc-proj.rs:21:45
   |
LL |     pub fn demo(_: impl Quux<Assoc=impl Foo<impl Bar>>) { }
   |                                    ---------^^^^^^^^-
   |                                    |        |
   |                                    |        nested `impl Trait` here
   |                                    outer `impl Trait`
   |
note: lint level defined here
  --> $DIR/issue-57979-nested-impl-trait-in-assoc-proj.rs:16:13
   |
LL |     #![warn(nested_impl_trait)]
   |             ^^^^^^^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #59014 <https://github.com/rust-lang/rust/issues/59014>

error: nested `impl Trait` is not allowed
  --> $DIR/issue-57979-nested-impl-trait-in-assoc-proj.rs:32:45
   |
LL |     pub fn demo(_: impl Quux<Assoc=impl Foo<impl Bar>>) { }
   |                                    ---------^^^^^^^^-
   |                                    |        |
   |                                    |        nested `impl Trait` here
   |                                    outer `impl Trait`
   |
note: lint level defined here
  --> $DIR/issue-57979-nested-impl-trait-in-assoc-proj.rs:27:13
   |
LL |     #![deny(nested_impl_trait)]
   |             ^^^^^^^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #59014 <https://github.com/rust-lang/rust/issues/59014>

error: aborting due to previous error

